### PR TITLE
Wallet Block Tools

### PR DESCRIPTION
**Pull request by Quexington**
This PR adds a new test class `WalletBlockTools`.  This class subclasses the standard `BlockTools` but overrides some methods around creating blocks.  The creates an approximate speed up of ~30% for the test runtimes in CI.

The motivation for this change is that wallet tests don't really depend on all of the consensus information being accurate for just testing normal wallet operations.  This consensus logic (while being something of a duplicated mess in `BlockTools` in the first place) is pretty slow and takes up test runtime for functionality that is not in any way being tested.

The `WalletBlockTools` class is used automatically by all tests in the `tests/wallet` directory.  There is an option to opt out of this behavior with the pytest mark `standard_block_tools`.  All blocks farmed with this class do the bare minimum calculation around transaction information and replaces the rest of the block record with dummy values like strings of `00` bytes or `None` etc.  To facilitate this, many methods in the full node and consensus directories have been monkey patched in the same fixture that patches the block tools.

The patching may seem somewhat disorganized, and this is because it's a result of empirical testing and patching until everything passed.  Nothing that has been patched is relevant to testing wallet operations, but there may be higher points to patch which make more sense.  It's my lack of overall understanding that requires me to go about it this way.  One day, it would be great to patch away the full node entirely to separate our project dependencies a little better.

There's also a decent number of fixes for tests in this PR that were needed due to new race conditions that were uncovered due to everything being faster/more syncronous.

### CODE REVIEW

File: chia/wallet/wallet_state_manager.py [hunk 1]
```diff
@@ -596,7 +596,9 @@ def tx_pending_changed(self) -> None:
 
         self.pending_tx_callback()
 
-    async def synced(self) -> bool:
+    async def synced(self, block_is_current_at: Optional[int] = None) -> bool:
```
Comment (arvidn) [inline chia/wallet/wallet_state_manager.py:599]: I can't find a place in this patch where `block_is_current_at` is passed in, when calling this function. Without a lot of context; it seems totally reasonable to be able to avoid have this function ask for what time it is. But it would seem simpler to just pass in the current time, rather than the time at which we consider blocks to be synced (i.e. 7 minutes ago, from the current time).

This same comment applies to `full_node.py` as well.

You could just take `now` as an optional parameter.

File: chia/simulator/full_node_simulator.py [hunk 2]
```diff
@@ -704,3 +723,14 @@ async def wait_for_wallets_synced(
                 if await self.wallets_are_synced(wallet_nodes=wallet_nodes):
                     break
                 await asyncio.sleep(backoff_time)
+
+    async def wait_for_self_synced(
+        self,
+        timeout: Optional[float] = 5,
```
Comment (arvidn) [inline chia/simulator/full_node_simulator.py:729]:
```suggestion
        timeout: float = 5,
```
`adjusted_timeout()` seems to take a `float`, not an `Optional[float]`

File: tests/wallet/conftest.py [hunk 3]
```diff
@@ -2,18 +2,85 @@
...
+from chia.wallet.wallet_state_manager import WalletStateManager
 from tests.environments.wallet import WalletEnvironment, WalletState, WalletTestFramework
 from tests.util.setup_nodes import setup_simulators_and_wallets_service
+from tests.wallet.wallet_block_tools import WalletBlockTools
+
+
+@pytest.fixture(scope="function", autouse=True)
+def block_is_current_at(monkeypatch: pytest.MonkeyPatch) -> None:
+    def make_new_synced(func: Callable[..., Awaitable[bool]]) -> Any:
+        async def mocked_synced(self: Any, block_is_current_at: int = 0) -> bool:
```
Comment (arvidn) [inline tests/wallet/conftest.py:27]:
```suggestion
        async def mocked_synced(self: Any, block_is_current_at: Optional[uint64] = 0) -> bool:
```

File: tests/wallet/conftest.py [hunk 4]
```diff
@@ -2,18 +2,85 @@
...
+    monkeypatch.setattr("chia.consensus.block_record.BlockRecord.sp_total_iters", lambda *_: uint128(0))
+    monkeypatch.setattr("chia.consensus.block_record.BlockRecord.ip_sub_slot_total_iters", lambda *_: uint128(0))
+    monkeypatch.setattr("chia.consensus.make_sub_epoch_summary.calculate_sp_iters", lambda *_: uint64(0))
+    monkeypatch.setattr("chia.consensus.make_sub_epoch_summary.calculate_ip_iters", lambda *_: uint64(0))
+    monkeypatch.setattr("chia.consensus.difficulty_adjustment._get_next_sub_slot_iters", lambda *_: uint64(1))
+    monkeypatch.setattr("chia.consensus.difficulty_adjustment._get_next_difficulty", lambda *_: uint64(1))
+    monkeypatch.setattr("chia.full_node.full_node_store.calculate_sp_interval_iters", lambda *_: uint64(1))
+    monkeypatch.setattr("chia.consensus.pot_iterations.calculate_sp_interval_iters", lambda *_: uint64(1))
+    monkeypatch.setattr("chia.consensus.pot_iterations.calculate_ip_iters", lambda *_: uint64(1))
+    monkeypatch.setattr("chia.consensus.block_record.BlockRecord.sp_sub_slot_total_iters", lambda *_: uint64(1))
```
Comment (arvidn) [inline tests/wallet/conftest.py:83]: is this a better approach than to make it possible to create a `BlockTool` object with flags (or perhaps just a `bool`) specifying what it needs to support and not. e.g. you could imagine creating a `BlockTools` in wallet-mode.

File: tests/wallet/conftest.py [hunk 5]
```diff
@@ -2,18 +2,85 @@
...
+            return await func(self, block_is_current_at)
+
+        return mocked_synced
+
+    monkeypatch.setattr(WalletStateManager, "synced", make_new_synced(WalletStateManager.synced))
+    monkeypatch.setattr(FullNode, "synced", make_new_synced(FullNode.synced))
+
+
+@pytest.fixture(scope="function", autouse=True)
+async def ignore_block_validation(request: pytest.FixtureRequest, monkeypatch: pytest.MonkeyPatch) -> None:
```
Comment (arvidn) [inline tests/wallet/conftest.py:37]: I can see, two years from now, somebody coming across this code and think: "I can simplify all this code by making the wallet tests use the same `BlockTools` as all the other tests.

I think there should be a comment here, explaining why that's a bad idea.

File: tests/wallet/test_singleton_lifecycle.py [hunk 6]
```diff
@@ -1,144 +0,0 @@
-from __future__ import annotations
-
-from typing import List, Tuple
```
Comment (arvidn) [inline tests/wallet/test_singleton_lifecycle.py:3]: did you mean to delete this? It seems unrelated to the main rationale for this change

**Review by arvidn — COMMENTED** [review]
> There's also a decent number of fixes for tests in this PR that were needed due to new race conditions that were uncovered due to everything being faster/more syncronous.

I think it would be great to pull those out and land them in a separate PR

> The patching may seem somewhat disorganized, and this is because it's a result of empirical testing and patching until everything passed

I'm a little bit worried that someone, in the future, will tear their hair out trying to understand why their shiny new wallet test is failing, when it's because some aspect of the consensus isn't maintained.

File: chia/wallet/wallet_state_manager.py (hunk 1 above)
Comment (Quexington) [reply chia/wallet/wallet_state_manager.py:599]: > I can't find a place in this patch where block_is_current_at is passed in, when calling this function

Yeah the patch adds the _ability_ to override the "When is the block current" behavior if desired.  Then in `conftest.py` we monkeypatch in a new default of `0` making all blocks current.  The implicit constants of 5 and 7 minutes were causing issues so this gives us the ability to patch those numbers.

> You could just take now as an optional parameter.

As stated above, the issue is not the "now" so much as the "since when".  This value is pretty handy if you view the function like `.am_i_syned_since(this_time):`.

File: chia/simulator/full_node_simulator.py (hunk 2 above)
Comment (Quexington) [reply chia/simulator/full_node_simulator.py:729]: Maybe your editor isn't handling `@overload` properly: https://github.com/Chia-Network/chia-blockchain/blob/main/chia/util/timing.py#L41

File: tests/wallet/test_singleton_lifecycle.py (hunk 6 above)
Comment (Quexington) [reply tests/wallet/test_singleton_lifecycle.py:3]: Yeah, the test was failing and when I took a look at it, I realized it was an unfinished, obsolete test whose functionality is covered elsewhere.  Probably should have been its own PR but I figured it was a simple enough deletion I could sneak it in, my bad.

File: tests/wallet/conftest.py (hunk 4 above)
Comment (Quexington) [reply tests/wallet/conftest.py:83]: That's what the patching of `BlockTools` with `WalletBlockTools` is for.  I didn't see any need to get granular with the stuff being mocked because I can't envision a scenario where the wallet cares about the stuff being mocked like VDF iterations etc.  The rest of this patching is to make the production full node code handle the mocked blocks.  Seemed like a simpler & much better alternative than piping a `test: bool = False` argument through all of our production code.

Comment (arvidn) [reply tests/wallet/conftest.py:94]: I see, this isn't done to affect the `BlockTools` behavior, it's to make the full node accept the invalid blocks.
Having a config option `disable_block_validation` might be reasonable though.

Comment (Quexington) [reply tests/wallet/conftest.py:94]: You mean in the full node?  That seems a bit sketchy to me but at the very least its out of the scope of the PR.

**Review by arvidn — APPROVED** [review]
all the monkey-patching makes me nervous, but I can't think of a much better (and practical) approach

File: tests/wallet/conftest.py (hunk 4 above)
Comment (pmaslana) [reply tests/wallet/conftest.py:94]: @Quexington Can this conversation be resolved now?

**Review by AmineKhaldi — APPROVED** [review]
